### PR TITLE
Make bad "rust-call" arguments no longer ICE

### LINKED ISSUE #22565: ICE when simplest rust-call method is called
Derivation of https://github.com/rust-lang/rust/issues/16039, which was closed because it failed to compile for other reasons:

``` rust
struct A;
impl A {
    extern "rust-call" fn b(i: i32) {}
}

fn main () {
    A::b(10);
}
```

```
error: internal compiler error: argument to function with "rust-call" ABI is neither a tuple nor unit
note: the compiler unexpectedly panicked. this is a bug.
note: we would appreciate a bug report: http://doc.rust-lang.org/complement-bugreport.html
note: run with `RUST_BACKTRACE=1` for a backtrace
thread 'rustc' panicked at 'Box<Any>', /home/rustbuild/src/rust-buildbot/slave/nightly-dist-rustc-linux/build/src/libsyntax/diagnostic.rs:189

stack backtrace:
   1:     0x7f3d5137aa60 - sys::backtrace::write::h17046f691c944fd6wlB
   2:     0x7f3d513a3650 - failure::on_fail::hc6163ed550dbd15faOK
   3:     0x7f3d512e9e10 - rt::unwind::begin_unwind_inner::hf466231ca0fbfbbdvsK
   4:     0x7f3d4e6569c0 - rt::unwind::begin_unwind::h15994696547183619422
   5:     0x7f3d4e657180 - diagnostic::Handler::bug::h1bc3ded4b9cf6157abF
   6:     0x7f3d4f359030 - session::Session::bug::h0d2283befd4ab8993ds
   7:     0x7f3d500a4bd0 - trans::type_of::type_of_rust_fn::h653f330b805e918apWo
   8:     0x7f3d500acac0 - trans::base::decl_rust_fn::h55f15a19d1270dfe9gs
   9:     0x7f3d500bfe70 - trans::base::register_fn::h09121dffeb6c811fDXu
  10:     0x7f3d500c1e10 - trans::base::register_method::h49e3ff79a84c5d89ZBv
  11:     0x7f3d4ffd5b30 - trans::base::get_item_val::hc88b3b1356c15164vkv
  12:     0x7f3d4ffd3410 - trans::base::trans_item::ha4ea07b91896bde9tPu
  13:     0x7f3d500c20f0 - trans::base::trans_crate::hc435c9fdb6f30ff4NMv
  14:     0x7f3d519c5250 - driver::phase_4_translate_to_llvm::hb02bd998f41b29a0wPa
  15:     0x7f3d5199e170 - driver::compile_input::h91ff1ed16e22d05cEba
  16:     0x7f3d51a71ea0 - run_compiler::h9c581f85e29a6b265bc
  17:     0x7f3d51a70500 - thunk::F.Invoke<A, R>::invoke::h5155567596601403298
  18:     0x7f3d51a6f3f0 - rt::unwind::try::try_fn::h5629523784837248188
  19:     0x7f3d5140ff20 - rust_try_inner
  20:     0x7f3d5140ff10 - rust_try
  21:     0x7f3d51a6f6a0 - thunk::F.Invoke<A, R>::invoke::h7807062886684583533
  22:     0x7f3d5138f190 - sys::thread::thread_start::h57c9c75c73c301cbfVF
  23:     0x7f3d4b3312b0 - start_thread
  24:     0x7f3d50f70249 - __clone
  25:                0x0 - <unknown>
```

### PR DESCRIPTION
**Pull request by CraftSpider**
The simplest of bad rust-call definitions will no longer cause an ICE. There is a FIXME added for future work, as I wanted to get this easy fix in before trying to either add a hack or mess with the whole obligation system

fixes #22565

### CODE REVIEW

File: compiler/rustc_typeck/src/check/mod.rs [hunk 1]
```diff
@@ -520,6 +521,35 @@ fn typeck_with_fallback<'tcx>(
 
             let fn_sig = fixup_opaque_types(tcx, &fn_sig);
 
+            if fn_sig.abi == abi::Abi::RustCall {
+                let expected_args = if let ImplicitSelfKind::None = decl.implicit_self { 1 } else { 2 };
+
+                let err = || {
+                    if let Node::Item(item) = tcx.hir().get(id) {
```
Comment (jyn514) [inline compiler/rustc_typeck/src/check/mod.rs:528]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/hir/map/struct.Map.html#method.expect_item
```suggestion
                    let item = tcx.hir().expect_item(id);
```

File: compiler/rustc_typeck/src/check/mod.rs [hunk 2]
```diff
@@ -520,6 +521,35 @@ fn typeck_with_fallback<'tcx>(
...
+                        }
+                    } else {
+                        bug!("Couldn't get span of FnHeader being checked")
+                    }
+                };
+
+                if fn_sig.inputs().len() != expected_args {
+                    err()
+                } else {
+                    match fn_sig.inputs()[expected_args - 1].kind() {
```
Comment (jyn514) [inline compiler/rustc_typeck/src/check/mod.rs:540]:
```suggestion
                    if !matches!(fn_sig.inputs()[expected_args - 1].kind(), ty::Tuple(_) | ty::Param(_)) {
```

**Comment by CraftSpider** [comment]
Note: overloaded-calls-nontuple is failing due to checking a case that is now being explicitly forbidden at the implementation location instead of later on. How should that be handled?

File: compiler/rustc_typeck/src/check/mod.rs [hunk 3]
```diff
@@ -520,6 +521,31 @@ fn typeck_with_fallback<'tcx>(
...
+                    }
+                };
+
+                if fn_sig.inputs().len() != expected_args {
+                    err()
+                } else {
+                    // FIXME(CraftSpider) Add a check on parameter expansion, so we don't just make the ICE happen later on
+                    //   This will probably require wide-scale changes to support a TupleKind obligation
+                    //   We can't resolve this without knowing the type of the param
+                    if !matches!(fn_sig.inputs()[expected_args - 1].kind(), ty::Tuple(_) | ty::Param(_)) {
```
Comment (oli-obk) [inline compiler/rustc_typeck/src/check/mod.rs:543]: won't this panic if you have a function with no arguments? Please add a test for that, too.

File: compiler/rustc_typeck/src/check/mod.rs [hunk 4]
```diff
@@ -520,6 +521,31 @@ fn typeck_with_fallback<'tcx>(
 
             let fn_sig = fixup_opaque_types(tcx, &fn_sig);
 
+            if fn_sig.abi == abi::Abi::RustCall {
```
Comment (oli-obk) [inline compiler/rustc_typeck/src/check/mod.rs:524]: I think this code should be in `check_fn`

**Comment by oli-obk** [comment]
> overloaded-calls-nontuple is failing due to checking a case that is now being explicitly forbidden at the implementation location instead of later on.

I'm not sure I understand, can you elaborate? I would have thought the failure is because you call `let item = tcx.hir().expect_item(id);` and methods are impl-items and not items.

**Comment by CraftSpider** [comment]
Nevermind, you're right, I misread the error. I'll fix that, and have a go at moving it into `check_fn`.

**Comment by CraftSpider** [comment]
Having fixed the ICE, now I'm getting the issue I mentioned, where the expected error is emitted, but so are 2 other errors saying `A function with the "rust-call" ABI must take a single non-self argument that is a tuple`. Is this valid to bless, as this is in fact an error as it stands? (Prior this commit, if you change the code to `S::call_once(s, 3)`, you get the error this is meant to fix, so it is a valid emission of this new error)

**Comment by oli-obk** [comment]
> Is this valid to bless, as this is in fact an error as it stands? (Prior this commit, if you change the code to S::call_once(s, 3), you get the error this is meant to fix, so it is a valid emission of this new error)

Yes, feel free to just bless all error changes. Especially after fixing an ICE this is very normal that more errors occur.

File: compiler/rustc_typeck/src/check/mod.rs (hunk 4 above)
Comment (CraftSpider) [reply compiler/rustc_typeck/src/check/mod.rs]: I moved it, I tried to put it near where I saw another ABI-related check was

**Pull request merged**